### ISSUE TITLE
Line breaking messing up when shortcode is used right before a line break

### ISSUE THREAD

**Issue by robertwbradford** [kernow/shortcode#37]
Nice gem! This will be a long, but hopefully helpful, description of the issue.

I have integrated this into a [Middleman](https://middlemanapp.com/) site and setup a [link] shortcode. However, I noticed when the [link] shortcode is at the end of a line right before the break, something is messing up.
# A little background...

Similar to Rails, Middleman uses (sometimes modified) versions of the [Padrino](http://www.padrinorb.com/guides/application-helpers) helpers. (Notably the `simple_format` method which is similar to Rails'.)

Also in Middleman, you can specify some [YAML Frontmatter](https://middlemanapp.com/basics/frontmatter/) which is then available in your page template.
# Now my setup

My [link] shortcode can accept an `id` attribute for referencing another page in the site. (Similar to the [RB Internal Links](https://wordpress.org/plugins/rb-internal-links/) WordPress plugin.) For example:

```
I want a [link id="123"]link to page with id of "123"[/link]. 
```

This would output the following:

```
I want a <a href="/the/path/to/page/123">link to page with id of "123"</a>.
```

In my page I have the following Frontmatter. Notice that the [link] shortcode occurs right before the two line breaks.

```

---
some_copy: |
  This is some copy. [link id="123"](And here is a link to page 123.)[/link]

  This is a second paragraph of copy.

---
```

I have a helper method for processing the shortcodes:

```
# compose_helpers.rb

def compose(something='')
  # I have a presenter for this shortcode that accepts the Middleman application object as the :app additional attribute. (Needed for handling the page lookup.)
  Shortcode.process(something, { app: self }).html_safe
end
```

Now in my layout for displaying this page I have the following:

```
<%= simple_format compose(current_page.data.some_copy) %>
```
# The problem

Normally the `simple_format` will convert two line breaks into `<p>` tags. However, something is happening when the shortcode is parsed and it looks like one of the line breaks disappears. `simple_format` then sees this as a single line break and uses a `<br>` instead of creating paragraphs.

In other words,

```
<%= simple_format current_page.data.some_copy %>
```

would normally output the following:

```
<p>This is some copy. [link id="123"](And here is a link to page 123.)[/link]</p>
<p>This is a second paragraph of copy.</p>
```

However, if I tell it to process the shortcodes first

```
<%= simple_format compose(current_page.data.some_copy) %>
```

I get the following:

```
<p>This is some copy. <a href="/path/to/page/123">(And here is a link to page 123.)</a><br>This is a second paragraph of copy.</p>
```

I'm not sure where the problem is. Any ideas?

**Comment by robertwbradford** [comment]
I noticed I if put a space character right after that closing `[/link]` then everything works fine. I've been looking at `lib/shortcode/parser.rb` but will need to study a little more to figure out what is going on.

**Comment by kernow** [comment]
@robertwbradford thanks for the detailed report. I'm away at the moment so won't be able to look at this until next week. Take a look at the tests, I think there was an issue with carriage returns right after a tag so there should be tests them, take a look and see if this covers your case and it not try creating a failing test with your use case. I can then look into it next week if you've not managed to work it out before then.

**Comment by robertwbradford** [comment]
Thanks, @kernow. I may have some time later today to go through the tests. If not, next week.

**Comment by kernow** [comment]
I've not been able to replicate this issue on the 2.0 branch, newlines are respected when parsing tags so I'm going to close this issue. If you are still experiencing this issue please reopen with a failing test example.
